### PR TITLE
add downstream local and remote address in StreamInfo interface for golang extension

### PR DESCRIPTION
**Pull request by StarryVae**
Commit Message: add downstream local and remote address in StreamInfo interface for golang extension
Additional Description: N/A
Risk Level: low
Testing: integration test
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
issue #26722

**Comment by StarryVae** [comment]
@doujiang24 PTAL, thanks.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 1]
```diff
@@ -904,6 +904,12 @@ CAPIStatus Filter::getStringValue(int id, GoString* value_str) {
     }
     req_->strValue = state.streamInfo().responseCodeDetails().value();
     break;
+  case EnvoyValue::DownstreamLocalAddress:
+    if (!state.streamInfo().downstreamAddressProvider().localAddress()) {
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:908]: seems the address must be there, we could simplify this API.
just like the Lua filter does: https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/lua/wrappers.cc#L190

File: contrib/golang/filters/http/source/go/pkg/api/filter.go [hunk 2]
```diff
@@ -69,6 +69,8 @@ type StreamInfo interface {
 	AttemptCount() uint32
 	// Get the dynamic metadata of the request
 	DynamicMetadata() DynamicMetadata
+	// DownstreamLocalAddress return the downstream local address.
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/api/filter.go:72]: does this API return the client address, or the server(envoy) address? I'm confused about it.
see: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/lua_filter#downstreamlocaladdress
> Returns the string representation of [downstream remote address](https://github.com/envoyproxy/envoy/blob/ea341eba3152aeda0ef0579bca61f5a1fe483203/envoy/stream_info/stream_info.h) used by the current request.

Comment (StarryVae) [reply contrib/golang/filters/http/source/go/pkg/api/filter.go:72]: it is the the server(envoy) address, maybe the lua doc is error, i will submit another pr to fix, thanks.

File: contrib/golang/filters/http/source/golang_filter.cc (hunk 1 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: done, thanks.

**Review by doujiang24 — COMMENTED** [review]
maybe we can implement DownstreamRemoteAddress here? another PR is also ok to me, just prefer here in personal.

otherwise, LGTM.

**Comment by StarryVae** [comment]
yeah, `DownstreamRemoteAddress` can also be implement in this PR.

**Comment by StarryVae** [comment]
> maybe we can implement DownstreamRemoteAddress here? another PR is also ok to me, just prefer here in personal.
> 
> otherwise, LGTM.

done, PTAL again, thanks

**Review by doujiang24 — APPROVED** [review]

**Comment by doujiang24** [comment]
/assign @wbpcode 
Thanks.

**Comment by StarryVae** [comment]
By the way, what about the `upstreamHost` and `upstreamCluster` api in StreamInfo?  It is also implement in WASM and our internal lua extension which may be also useful. cc @doujiang24

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by doujiang24** [comment]
> By the way, what about the `upstreamHost` and `upstreamCluster` api in StreamInfo? It is also implement in WASM and our internal lua extension which may be also useful. cc @doujiang24

yep, they could be useful, PR welcome~

**Pull request merged**